onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -120,35 +120,72 @@ bool ReshapeFusion::Fuse_Subgraph1(Node& reshape, Graph& graph, const logging::L
     return false;
   }
 
-  // Loop through the inputs of concat node to calculate the shape_value for a potential reshape fusion. 
+  // Loop through the inputs of concat node to calculate the shape_value for a potential reshape fusion.
   std::vector<int64_t> shape_value;
   shape_value.reserve(concat_input_count);
   for (int i = 0; i < concat_input_count; ++i) {
     // First check if the i-th argument is a constant initializer.
     if (optimizer_utils::AppendTensorFromInitializer(graph, *(concat.InputDefs()[i]), shape_value, true)) {
       continue;
     }
+
     // Try to find path [Root] --> Shape --> Gather(indices=i) --> Unsqueeze (axes=0) --> Concat [input i]
     bool matched = ReshapeFusion::Fuse_Subgraph2(graph, root_input, concat, i, shape_value, logger);
-    const Node* p_cur_node = graph_utils::GetInputNode(concat, i);
     if (matched) {
       shape_value.push_back(0);
-    } else if (p_cur_node != nullptr) {
-      // This node could lead to a potential subgraph pattern fusion. Mark the shape value to -1.
-      shape_value.push_back(-1);
-    } else {
+      // We have matched the pattern for this input into Concat
+      // Proceed to the next input
+      continue;
+    }
+
+    // If we haven't been able to match the pattern, try and see if this is a candidate for subgraph pattern fusion.
+    // For this input to be a candidate, the number of elements in the input tensor to Concat has to be 1
+    // We use shape info (if made available via shape inference) for this.
+    const NodeArg* concat_input_node_arg = concat.InputDefs()[i];
+    if (!concat_input_node_arg) {
+      // We need NodeArg to be able to lookup any shape info
+      // Can't proceed with fusion
       return false;
     }
-  }  
 
-  // Check how many -1 are there in shape_value.
+    // ONNX shape inference runs after Level 1 optimization and hence shape information may not be available
+    // at this point.
+    // The shape inference might be available if the shape info was available via the `value_info` field in GraphProto
+    // or if the model had shape inference run on it which made the shape information available
+
+    const auto* input_shape = concat_input_node_arg->Shape();
+    if (!input_shape) {
+      // We need shape to be able to be certain of number of elements
+      // Can't proceed with fusion
+      return false;
+    }
+
+    // For the number of elements in this input to Concat be 1,
+    // all dim values in the shape must be 1.
+    const auto input_rank = concat_input_node_arg->Shape()->dim_size();
+
+    for (int dim_index = 0; dim_index < input_rank; ++dim_index) {
+      const auto& dim = input_shape->dim()[dim_index];
+      if (!dim.has_dim_value() || dim.dim_value() != 1) {
```

Comment:
Basically this is a size check. If the tensor feeding into Concat has one element, we may proceed with fusion